File: front/src/lib/schemas/technical-survey.ts
import { z } from "zod";

/**
 * Schéma de base pour Technical Survey.
 * Tous les champs sont optionnels pour permettre les sauvegardes partielles (draft).
 */
export const technicalSurveyBaseSchema = z.object({
  photo_house: z.string().optional().nullable(),
  photo_facade: z.string().optional().nullable(),
  photo_old_system: z.string().optional().nullable(),
  photo_electric_panel: z.string().optional().nullable(),
  has_linky: z.boolean().optional().nullable(),
  photo_linky: z.string().optional().nullable(),
  photo_breaker: z.string().optional().nullable(),
});

export type TechnicalSurveyBaseValues = z.infer<typeof technicalSurveyBaseSchema>;

/**
 * Schéma pour sauvegarde complète (tous les champs requis).
 * Utilise superRefine pour la logique conditionnelle Linky.
 */
export const technicalSurveyCompleteSchema = technicalSurveyBaseSchema
  .superRefine((data, ctx) => {
    // Les 4 photos standard sont toujours requises
    if (!data.photo_house) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "La photo du logement est requise",
        path: ["photo_house"],
      });
    }
    if (!data.photo_facade) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "La photo de la façade est requise",
        path: ["photo_facade"],
      });
    }
    if (!data.photo_old_system) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "La photo de l'ancien système est requise",
        path: ["photo_old_system"],
      });
    }
    if (!data.photo_electric_panel) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "La photo du tableau électrique est requise",
        path: ["photo_electric_panel"],
      });
    }

    // Logique conditionnelle Linky
    if (data.has_linky === true) {
      if (!data.photo_linky) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "La photo du compteur Linky est requise",
          path: ["photo_linky"],
        });
      }
    } else if (data.has_linky === false) {
      if (!data.photo_breaker) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "La photo du disjoncteur est requise",
          path: ["photo_breaker"],
        });
      }
    } else {
      // has_linky n'est pas défini
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "Veuillez indiquer si le logement a un compteur Linky",
        path: ["has_linky"],
      });
    }
  });

export type TechnicalSurveyCompleteValues = z.infer<
  typeof technicalSurveyCompleteSchema
>;

/**
 * Schéma pour sauvegarde draft (validation assouplie).
 * Seule la logique conditionnelle Linky est appliquée.
 */
export const technicalSurveyDraftSchema = technicalSurveyBaseSchema.superRefine(
  (data, ctx) => {
    // Logique conditionnelle Linky uniquement
    if (data.has_linky === true && !data.photo_linky) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "La photo du compteur Linky est requise",
        path: ["photo_linky"],
      });
    } else if (data.has_linky === false && !data.photo_breaker) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "La photo du disjoncteur est requise",
        path: ["photo_breaker"],
      });
    }
  }
);

export type TechnicalSurveyDraftValues = z.infer<
  typeof technicalSurveyDraftSchema
>;

File: back/app/schemas/technical_survey.py
"""
Pydantic schemas for Technical Survey.
"""
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class TechnicalSurveyBase(BaseModel):
    """Champs communs pour les photos de visite technique."""

    photo_house: str | None = Field(
        None,
        max_length=512,
        description="Photo du logement.",
    )
    photo_facade: str | None = Field(
        None,
        max_length=512,
        description="Photo de la façade.",
    )
    photo_old_system: str | None = Field(
        None,
        max_length=512,
        description="Photo de l'ancien système (chaudière).",
    )
    photo_electric_panel: str | None = Field(
        None,
        max_length=512,
        description="Photo du tableau électrique.",
    )
    has_linky: bool | None = Field(
        None,
        description="Le logement a-t-il un compteur Linky ?",
    )
    photo_linky: str | None = Field(
        None,
        max_length=512,
        description="Photo du compteur Linky (si has_linky=True).",
    )
    photo_breaker: str | None = Field(
        None,
        max_length=512,
        description="Photo du disjoncteur (si has_linky=False).",
    )


class TechnicalSurveyCreate(TechnicalSurveyBase):
    """Payload de création des photos de visite technique."""

    pass


class TechnicalSurveyUpdate(BaseModel):
    """Payload de mise à jour partielle des photos de visite technique."""

    photo_house: str | None = Field(
        None,
        max_length=512,
        description="Photo du logement.",
    )
    photo_facade: str | None = Field(
        None,
        max_length=512,
        description="Photo de la façade.",
    )
    photo_old_system: str | None = Field(
        None,
        max_length=512,
        description="Photo de l'ancien système (chaudière).",
    )
    photo_electric_panel: str | None = Field(
        None,
        max_length=512,
        description="Photo du tableau électrique.",
    )
    has_linky: bool | None = Field(
        None,
        description="Le logement a-t-il un compteur Linky ?",
    )
    photo_linky: str | None = Field(
        None,
        max_length=512,
        description="Photo du compteur Linky (si has_linky=True).",
    )
    photo_breaker: str | None = Field(
        None,
        max_length=512,
        description="Photo du disjoncteur (si has_linky=False).",
    )


class TechnicalSurveyResponse(TechnicalSurveyBase):
    """Représentation API des photos de visite technique."""

    folder_id: UUID
    tenant_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
